pkg/controller/networkpolicy/crd_utils.go

```diff
@@ -42,18 +46,96 @@ func toAntreaServicesForCRD(npPorts []secv1alpha1.NetworkPolicyPort) ([]controlp
 	var antreaServices []controlplane.Service
 	var namedPortExists bool
 	for _, npPort := range npPorts {
-		if npPort.Port != nil && npPort.Port.Type == intstr.String {
-			namedPortExists = true
+		if npPort.Port != nil {
+			namedPortExists = addSingleAntreaServiceForCRD(npPort.Protocol, npPort.Port, &antreaServices)
 		}
-		antreaService := controlplane.Service{
-			Protocol: toAntreaProtocol(npPort.Protocol),
-			Port:     npPort.Port,
+		if npPort.PortRange != nil {
+			if npPort.PortRange.Port != nil {
+				namedPortExists = addSingleAntreaServiceForCRD(npPort.Protocol, npPort.PortRange.Port, &antreaServices) || namedPortExists
+			} else if npPort.PortRange.From != nil && npPort.PortRange.To != nil {
+				if npPort.PortRange.Except != nil {
+					portRanges := rangeSplit(*npPort.PortRange.From, *npPort.PortRange.To, npPort.PortRange.Except)
+					for _, portRange := range portRanges {
+						err := addRangeAntreaServiceForCRD(npPort.Protocol, portRange, &antreaServices)
+						if err != nil {
+							klog.Error(err.Error())
+						}
+					}
+				} else {
+					err := addRangeAntreaServiceForCRD(npPort.Protocol, network_policy.PortRange{Start: *npPort.PortRange.From, End: *npPort.PortRange.To}, &antreaServices)
+					if err != nil {
+						klog.Error(err.Error())
+					}
+				}
+			}
+		}
+		if npPort.Port == nil && npPort.PortRange == nil {
+			antreaServices = append(antreaServices, controlplane.Service{Protocol: toAntreaProtocol(npPort.Protocol)})
 		}
-		antreaServices = append(antreaServices, antreaService)
 	}
 	return antreaServices, namedPortExists
 }
 
+// Add single antrea service with single port
+func addSingleAntreaServiceForCRD(protocol *v1.Protocol, port *intstr.IntOrString, antreaServices *[]controlplane.Service) bool {
```

Comment:
slice is already a pointer internally, no need for the pointer around `[]controlplane.Service`

Reply:
If you use `append` to modify the slice inside a function, it will be different from your thought. https://stackoverflow.com/questions/49428716/pass-slice-as-function-argument-and-modify-the-original-slice.
If you think using the pointer of a slice is too tricky I can change to make that function to return the slice.